### ISSUE TITLE
Enable utsname on MacOS

### ISSUE THREAD

**Issue by zsck** [nix-rust/nix#607]
I'm doing a really quick experiment to try to invoke the `uname` function on macOS. Following [the documentation](https://docs.rs/nix/0.8.1/nix/sys/utsname/fn.uname.html), I wrote the code below.

```rust
extern crate nix;

fn main() {
    println!("Hello, world!");
    let name = nix::sys::utsname::uname();
    println!("
      sysname = {}
      nodename = {}
      release = {}
      version = {}
      machine = {}
    ", name.sysname(), name.nodename(), name.release(), name.version(), name.machine());
}
```

My `Cargo.toml` file points to version 0.8.1 of `nix` as a dependency.  When I try to compile this program with `cargo build` I get the following error:

```
error[E0433]: failed to resolve. Could not find `utsname` in `sys`
 --> src/main.rs:5:16
  |
5 |     let name = nix::sys::utsname::uname();
  |                ^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `utsname` in `sys`
```

I'm running this on macOS 10.12.4 with rustc 1.17.0.

**Comment by zsck** [comment]
I see that `utsname` is actually only exported [for Linux and Android](https://github.com/nix-rust/nix/blob/master/src/sys/mod.rs#L45) targets.  Would it not work on macOS?

**Comment by asomers** [comment]
@zsck try it and find out.  If it does, submit a PR to enable it.  Include in the PR a link to the relevant man page.

**Comment by zsck** [comment]
I modified `src/sys/mod.rs` to export `utsname` on macOS like so:

```rust
#[cfg(any(target_os = "linux", target_os = "android", target_os = "macos"))]
pub mod utsname;
```

Now my program compiles and runs, however it terminates silently as soon as `uname` is called. The following `println!` isn't executed.

Unfortunately I am not equipped to figure out how to resolve this issue.

**Comment by zsck** [comment]
Interestingly, if I modify the wrapper for `uname` so that it looks like so:

```rust
pub fn uname() -> UtsName {
    unsafe {
        let mut ret: UtsName = mem::uninitialized();
        let retcode = ffi::uname(&mut ret as *mut UtsName);
        println!("RETURN CODE: {}", retcode);
        ret
    }
}
```

Then re-run my program, I now see the following output:

```
Hello, world!
RETURN CODE FROM uname: 0

      sysname = Darwin
      nodename = R�
      release =
      version =
      machine =
```

The value of `nodename` appears to change every time I run my program, and sometimes there are extra whitespace characters in the value of `version`.

**Comment by Susurrus** [comment]
@zsck Just wanted to point out that your documentation link is incorrect. It points to the docs for Linux (which is why that function exists). You really want to explore the docs [here](https://docs.rs/nix/0.8.1/x86_64-apple-darwin/nix/). Note the dropdown to select the Platform at the top of the page. This is common with multi-platform crates that have platform-specific functionality. Be sure to read the right docs!

As to the actual issue, the proper place to bring this up actually isn't here, but in [libc](https://github.com/rust-lang/libc). As you've seen, nix just provides a safe wrapper around `uname`, but the FFI is done by the `libc` crate. Please open an issue there about this (and be sure to link to this issue from there!). Once that's resolved we can expose it on MacOS if appropriate.

**Comment by zsck** [comment]
I am not 100% sure if this is in fact a problem with libc.  I wrote the following code using libc, and it seems to work just fine.

```rust
extern crate libc;

use std::mem;

fn main() {
    println!("Getting name");
    unsafe {
        let mut name: libc::utsname = mem::uninitialized();
        let ret_code = libc::uname(&mut name as *mut libc::utsname);
        println!("RETCODE = {}", ret_code);
        print!("version = ");
        for i in 0..name.version.len() {
            print!("{}", name.version[i] as u8 as char);
        }
    }
}
```

**Comment by Susurrus** [comment]
You didn't need to write code, just look at libc see and see where it's declared. What platforms do they support using it on?

**Comment by zsck** [comment]
I don't see anything indicating that `uname` wouldn't be available on both macOS and Linux.  Going through their documentation into the source, it appears [there are no restrictions](https://doc.rust-lang.org/libc/x86_64-apple-darwin/src/libc/unix/mod.rs.html#724) on which platforms `uname` is exported for.

I don't see how it's a problem of libc's if I'm able to use `uname` perfectly fine via libc, but I cannot using nix.

**Comment by asomers** [comment]
There seems to be an assumption that `nix::sys::utsname` has the same memory layout as `libc::utsname`.  That assumption is probably true on Linux but not OSX.  That's the reason that your first experiment didn't work.

**Comment by Susurrus** [comment]
@somers You're probably right that the memory layout is incorrect in `libc`.

First off you can see that `libc` defines `uname()` for all unix platforms (@zsck it's easiest just to look through the hierarchy in the `libc` source like [here](https://github.com/rust-lang/libc/blob/fda4049c11e75871a9e583fa5eea55f9be0f9972/src/unix/mod.rs#L724)). But `utsname` is defined per-platform ([see here](https://github.com/rust-lang/libc/search?utf8=%E2%9C%93&q=utsname&type=)). And you can see that the OS X implements is [here](https://github.com/rust-lang/libc/blob/5257dd8cc09df7d7763e6921de75cfcab47d7b3c/src/unix/bsd/mod.rs#L87) and that it's common among all BSDs that aren't Dragonfly.

**Comment by asomers** [comment]
@Susurrus My concern isn't that libc's definition for OSX's `utsname` is wrong, though that may be a problem too.  My main concern is that nix's `uname` passes a pointer to nix's `UtsName` structure directly to an FFI `uname` function.  The only way that can work is if nix's `Utsname` structure has the same layout as the structure that the OS's `uname` function expects.  That's very unlikely to be true on all platforms.
https://github.com/nix-rust/nix/blob/master/src/sys/utsname.rs#L59

**Comment by Susurrus** [comment]
Oh, man, our utsname code is completely wrong. We shouldn't have FFI definitions in our code. And really we should have a newtype and then pass that to the FFI function. As long as we specify it's `repr(C)` we should be fine.

**Comment by Susurrus** [comment]
Sorry, not completely wrong, but definitely not well written.

**Comment by Susurrus** [comment]
@zsck Would you be interested in filing a PR for this? You'd instead create a newtype for `UtsName` that looks like of like the `PtyMaster` struct I added in #556. Then defer all FFI to the `libc` crate. That will likely fix the issue.

**Comment by zsck** [comment]
I wish I could say that I would like to take care of this, but my use case for `uname` is very limited in scope and doesn't add a whole lot of value to the project I'm working on (for my employer).

**Comment by Susurrus** [comment]
This should be addressed by #739. @zsck please test out that PR and let me know if it works for you.
